src/openvic-simulation/history/ProvinceHistory.cpp

```diff
@@ -98,7 +102,16 @@ bool ProvinceHistoryMap::_load_history_entry(
 			expect_identifier(expect_mapped_string(colony_status_map, assign_variable_callback(entry.colonial))),
 		"is_slave", ZERO_OR_ONE, expect_bool(assign_variable_callback(entry.slave)),
 		"trade_goods", ZERO_OR_ONE,
-			good_definition_manager.expect_good_definition_identifier(assign_variable_callback_pointer_opt(entry.rgo)),
+			good_definition_manager.expect_good_definition_identifier_or_string(
+				[&definition_manager, &entry](GoodDefinition const& rgo_good) ->bool {
+					entry.rgo_production_type = definition_manager.get_economy_manager().get_production_type_manager().get_good_to_rgo_production_type()[rgo_good];
+					return entry.rgo_production_type != nullptr;
+					//we expect the good to have an rgo production type
+					//Victoria 2 treats this as null, but clearly the modder wanted there to be a good
+				},
+				allow_empty_true, //could be explicitly setting trade_goods to null
+				do_warn //could be typo in good identifier
```

Comment:
If there is no trade_goods entry, okay.
If there is a trade_goods entry with invalid goods identifier, treat as null + warning.
If there is a trade_goods entry with valid goods identifier that doesn't have rgo production type, error.
If there is a trade_goods entry with valid goods identifier that does have rgo production type, okay + change rgo.

Reply:
- `ZERO_OR_ONE` means no `trade_goods` entry is fine
- `allow_empty_true` also means `trade_goods = ""` produces no error and doesn't call the callback at all
- any non-empty string which isn't a valid good identifier won't call the callback either, and because `do_warn` is true this will produce a warning rather than an error
- a valid trade good with no production type will indeed produce an error, as the callback will be called and the `if` statement within it will be triggered
- a valid trade good with a non-null production type will work correctly, setting the rgo production type and not producing an error

So compared to what you state above, the only discrepancy is with the invalid good identifier case: an invalid trade good won't be treated as null, it will have no effect on `entry.rgo_production_type_nullable` at all (the callback can't be called unless there's a valid good to use for its argument). If the invalid trade good is due to an empty string then there'll be no warning, if it's from a non-empty string then there will be a warning. If you want a warning from empty strings too then replace `allow_empty_true` with a false value. If you want to force `entry.rgo_production_type_nullable` to be null when given an invalid trade good string (empty or not) then you'll need to use a regular `expect_identifier_or_string` and try find a `GoodDefinition` with that identifier in the callback, either prodeeding as normal if you find it or manually setting `entry.rgo_production_type_nullable` to `nullptr` if you don't.

Reply:
It works as intended then. If someone wants to use an invalid non-empty identifier for null and we somehow think that is fine, we can always change it later.